src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -587,10 +626,26 @@ private async Task InitializeAsync(bool ignoreFailures, IEnumerable<Configuratio
 
                 try
                 {
-                    Dictionary<string, ConfigurationSetting> mappedData = await MapConfigurationSettings(data).ConfigureAwait(false);
-                    SetData(await PrepareData(mappedData, cancellationToken).ConfigureAwait(false));
+                    _mappedData = await MapConfigurationSettings(data).ConfigureAwait(false);
                     _watchedSettings = watchedSettings;
-                    _mappedData = mappedData;
+                    Dictionary<string, ConfigurationSetting> orderedData = new Dictionary<string, ConfigurationSetting>();
+
+                    foreach (KeyValuePair<KeyValueIdentifier, ConfigurationSetting> kvp in _mappedData)
+                    {
+                        orderedData[kvp.Key.Key] = kvp.Value;
+                    }
+
+                    foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers.Concat(_options.MultiKeyWatchers))
+                    {
+                        IEnumerable<ConfigurationSetting> currentKeyValues = GetCurrentKeyValueCollection(changeWatcher.Key, changeWatcher.Label, _watchedSettings.Values);
+
+                        foreach (ConfigurationSetting setting in currentKeyValues)
+                        {
+                            orderedData[setting.Key] = _mappedData[new KeyValueIdentifier(setting.Key, setting.Label)];
+                        }
+                    }
```

Comment:
During initialization, we don't need to reorder watched settings because we just finished reloading the entire configuration. `_mappedData` should already be in the right order of precedence.